crates/bdk/src/wallet/mod.rs

```diff
@@ -2558,22 +2531,31 @@ fn create_signers<E: IntoWalletDescriptor>(
     index: &mut KeychainTxOutIndex<KeychainKind>,
     secp: &Secp256k1<All>,
     descriptor: E,
-    change_descriptor: Option<E>,
+    change_descriptor: E,
     network: Network,
-) -> Result<(Arc<SignersContainer>, Arc<SignersContainer>), crate::descriptor::error::Error> {
-    let (descriptor, keymap) = into_wallet_descriptor_checked(descriptor, secp, network)?;
+) -> Result<(Arc<SignersContainer>, Arc<SignersContainer>), DescriptorError> {
+    let descriptor = into_wallet_descriptor_checked(descriptor, secp, network)?;
+    let change_descriptor = into_wallet_descriptor_checked(change_descriptor, secp, network)?;
+    if descriptor.0 == change_descriptor.0 {
+        return Err(DescriptorError::NonUnique(descriptor.0.to_string()));
+    }
+
+    let (descriptor, keymap) = descriptor;
     let signers = Arc::new(SignersContainer::build(keymap, &descriptor, secp));
     index.add_keychain(KeychainKind::External, descriptor);
 
-    let change_signers = match change_descriptor {
-        Some(descriptor) => {
-            let (descriptor, keymap) = into_wallet_descriptor_checked(descriptor, secp, network)?;
-            let signers = Arc::new(SignersContainer::build(keymap, &descriptor, secp));
-            index.add_keychain(KeychainKind::Internal, descriptor);
-            signers
-        }
-        None => Arc::new(SignersContainer::new()),
-    };
+    let (descriptor, keymap) = change_descriptor;
+    let change_signers = Arc::new(SignersContainer::build(keymap, &descriptor, secp));
+    index.add_keychain(KeychainKind::Internal, descriptor);
+
+    // Check we didn't try to add an overlapping keychain
+    if index.spk_at_index(KeychainKind::Internal, 0).is_none() {
+        let change_descriptor = index
+            .keychains()
+            .get(&KeychainKind::Internal)
+            .expect("we added the keychain");
+        return Err(DescriptorError::NonUnique(change_descriptor.to_string()));
+    }
```

Comment:
you're right, it won't catch overlap if the non-wildcard descriptor has a derivation index greater than `DEFAULT_LOOKAHEAD`, so that's a problem.

Reply:
We could derive the spk for the non-wildcard descriptor, and then do an exhaustive search in the wildcard descriptor for an index matching that spk. It would of course be more efficient if we could infer the definite derivation index.

Reply:
I agree with this. We don't need to provide guarantees here. We can check equality of descriptors but other than that it's up to the programmer not to introduce "overlapping" descriptors.